Integrate NewFile Action

valentinpurrucker opened the pull request:
Closes #33 

- User can create a new file either from ContextMenu or from theToolbar
- TODO: An icon for the action NewFile

valentinpurrucker committed:
Add new file wizard and a new context menu child for this action.

Task: #33

valentinpurrucker committed:
Merge branch 'team02-development' into feature/task_33-new_file_action

Conflicts:
	de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java

Task: #33

timbittner commented:
Regex Bugs:
- The Regex does not allow:
  - Filenames starting with '.' -> ".gitignore"
  - Filenames containing valid special characters like "testv2.xml"
  - A few more but my PC crashed and I'm writing this the second time and am too lazy to test everything again

Is there maybe a method provided by the Java files interface, that checks if a given name is valid on an operating system? Who knows what will happen on linux distros now 😆 

Alternatively we can accept the fact that our file generation is always limited to this regex.

timbittner commented:
The in-wizard "Change Directory" Button allows for the target to be outside the scope of the workspace, which does not seem too bad but considering how mental average users are we can assume they will trip and fall over this and then complain that the feature does not work.

We can either catch that or remove the extra convenience alltogether with that button (other wizards don't have it either). Im open for discussion.

timbittner committed:
Add New-File/-Folder Icons and change Toolbar Icon order

New Folder and New File Icons are now available and replace their bland
counterparts where fit.
The Button sequence in Toolbar is now the same as in SWT context menu.
Also Updated 'Input Files' Tooltips and Labels to 'Import ...'

#37

timbittner commented:
> Closes #33
> 
> * User can create a new file either from ContextMenu or from theToolbar
> * TODO: An icon for the action NewFile

TODO Done

valentinpurrucker commented:
> The in-wizard "Change Directory" Button allows for the target to be outside the scope of the workspace, which does not seem too bad but considering how mental average users are we can assume they will trip and fall over this and then complain that the feature does not work.
> 
> We can either catch that or remove the extra convenience alltogether with that button (other wizards don't have it either). Im open for discussion.

Yeah if we add this behavior to this wizard we need to add it to others as well. I think it actually would be better to remove it

valentinpurrucker commented:
> Regex Bugs:
> 
> * The Regex does not allow:
>   
>   * Filenames starting with '.' -> ".gitignore"
>   * Filenames containing valid special characters like "testv2.xml"
>   * A few more but my PC crashed and I'm writing this the second time and am too lazy to test everything again
> 
> Is there maybe a method provided by the Java files interface, that checks if a given name is valid on an operating system? Who knows what will happen on linux distros now 😆
> 
> Alternatively we can accept the fact that our file generation is always limited to this regex.

One way we could do it is to rely on the Java `Paths.get(path: String)` method. This throws an `InvalidPathException` if the path cannot be converted. This solution also allows really weird names that include _%,&,$,§_ but we could argue that if java is fine with that, why not ?!. On the other hand the problem is with _.gitignore_  that the explorer does not show hidden files i guess, so the user does not see these files.

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/project_explorer.e4xmi:
```diff
@@ -4,8 +4,9 @@
       <menus xsi:type="menu:PopupMenu" xmi:id="_RuDIlffoEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.e4cf.parts.project_explorer.context_menu" label="project_explorer_context_menu">
-        <children xsi:type="menu:DirectMenuItem" xmi:id="_RuDIlvfoEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.familymining.ppu_iec.rcp_e4.directmenuitem.new_folder" label="New Folder" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/folder16.png" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.project_explorer/de.tu_bs.cs.isf.e4cf.parts.project_explorer.handl ...
-        <children xsi:type="menu:DirectMenuItem" xmi:id="_RuDIl_foEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.familymining.ppu_iec.rcp_e4.directmenuitem.import_files" label="Import Files" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/file16.png" tooltip="Import one or more files into a selected folder" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.projec ...
+        <children xsi:type="menu:DirectMenuItem" xmi:id="_RuDIlvfoEeiagocgPMLR0w" elementId="de.tu_bs.cs.isf.familymining.ppu_iec.rcp_e4.directmenuitem.new_folder" label="New Folder" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/newfolder16.png" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.project_explorer/de.tu_bs.cs.isf.e4cf.parts.project_explorer.ha ...
+        <children xsi:type="menu:DirectMenuItem" xmi:id="_76VKEEnZEeu4Feb2HZAxiA" elementId="de.tu_bs.cs.isf.e4cf.parts.project_explorer.directmenuitem.addnewfile" label="New File" iconURI="platform:/plugin/de.tu_bs.cs.isf.e4cf.core/icons/Explorer_View/items/newfile16.png" contributionURI="bundleclass://de.tu_bs.cs.isf.e4cf.parts.project_explorer/de.tu_bs.cs.isf.e4cf.parts.project_explorer.handle ...
```
Nice icons :)

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java:
```diff
@@ -6,13 +6,17 @@
 	public static final String PROJECT_EXPLORER_VIEW_FXML = Paths.get("ui", "view", "ProjectExplorerView.fxml")
 			.toString();
 
+	public static final String NEW_FILE_PAGE_VIEW = Paths.get("ui", "view", "NewFileView.fxml").toString();
```
Some nitpicking here: maybe a constant for ui/view?

valentinpurrucker replied:
makes sense

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/handlers/NewFileHandler.java:
```diff
@@ -0,0 +1,64 @@
+			directory = FileHandlingUtility.getPath(fileSystem.getWorkspaceDirectory());
+		}
+
+		if (context != null) {
+			this.context = context;
+		}
```
I dont get this: why do we field `@Inject` the context, also pass it and then set it here again? We only use it in this function, right?

valentinpurrucker replied:
We open the dialog both from a context menu and from code(from the toolbar). The problem is that when i try open it from the toolbar via a button it does not inject the context :D Don't know if we can do something about that

Jooshua reviewed (commented):
Good job!

Jooshua commented:
> 
> 
> > Regex Bugs:
> > 
> > * The Regex does not allow:
> >   
> >   * Filenames starting with '.' -> ".gitignore"
> >   * Filenames containing valid special characters like "testv2.xml"
> >   * A few more but my PC crashed and I'm writing this the second time and am too lazy to test everything again
> > 
> > Is there maybe a method provided by the Java files interface, that checks if a given name is valid on an operating system? Who knows what will happen on linux distros now 😆
> > Alternatively we can accept the fact that our file generation is always limited to this regex.
> 
> One way we could do it is to rely on the Java `Paths.get(path: String)` method. This throws an `InvalidPathException` if the path cannot be converted. This solution also allows really weird names that include _%,&,$,§_ but we could argue that if java is fine with that, why not ?!. On the other hand the problem is with _.gitignore_ that the explorer does not show hidden files i guess, so the user does not see these files.

Not sure if I get this right, but is showing hidden files not a windows explorer setting?

valentinpurrucker commented:
> > > Regex Bugs:
> > > 
> > > * The Regex does not allow:
> > >   
> > >   * Filenames starting with '.' -> ".gitignore"
> > >   * Filenames containing valid special characters like "testv2.xml"
> > >   * A few more but my PC crashed and I'm writing this the second time and am too lazy to test everything again
> > > 
> > > Is there maybe a method provided by the Java files interface, that checks if a given name is valid on an operating system? Who knows what will happen on linux distros now 😆
> > > Alternatively we can accept the fact that our file generation is always limited to this regex.
> > 
> > 
> > One way we could do it is to rely on the Java `Paths.get(path: String)` method. This throws an `InvalidPathException` if the path cannot be converted. This solution also allows really weird names that include _%,&,$,§_ but we could argue that if java is fine with that, why not ?!. On the other hand the problem is with _.gitignore_ that the explorer does not show hidden files i guess, so the user does not see these files.
> 
> Not sure if I get this right, but is showing hidden files not a windows explorer setting?

@Jooshua  as far as i can tell i can see the hidden files in my explorer (e.g. the .gitignore file) but not in the project explorer

valentinpurrucker committed:
Changes due to PR

- remove directory chooser
- small changes

Task: #33

kyshidar commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/.settings/org.eclipse.core.resources.prefs:
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
+encoding//src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/wizards/NewFilePage.java=UTF-8
```
why does  the encoding have to specified for this one file explicitly? Just wondering.

valentinpurrucker replied:
i did not understand that too. it has to do with the Regex, because when i first implemented it it gave me an error that we need utf8 encoding

kyshidar replied:
okay then

timbittner commented:
> > > > Regex Bugs:
> > > > 
> > > > * The Regex does not allow:
> > > >   
> > > >   * Filenames starting with '.' -> ".gitignore"
> > > >   * Filenames containing valid special characters like "testv2.xml"
> > > >   * A few more but my PC crashed and I'm writing this the second time and am too lazy to test everything again
> > > > 
> > > > Is there maybe a method provided by the Java files interface, that checks if a given name is valid on an operating system? Who knows what will happen on linux distros now 😆
> > > > Alternatively we can accept the fact that our file generation is always limited to this regex.
> > > 
> > > 
> > > One way we could do it is to rely on the Java `Paths.get(path: String)` method. This throws an `InvalidPathException` if the path cannot be converted. This solution also allows really weird names that include _%,&,$,§_ but we could argue that if java is fine with that, why not ?!. On the other hand the problem is with _.gitignore_ that the explorer does not show hidden files i guess, so the user does not see these files.
> > 
> > 
> > Not sure if I get this right, but is showing hidden files not a windows explorer setting?
> 
> @Jooshua as far as i can tell i can see the hidden files in my explorer (e.g. the .gitignore file) but not in the project explorer

Same for me. So then let's ignore hidden files / files starting with . and call it a bug if @KamilRosiak even wants to support this feature :)

valentinpurrucker commented:
> > > > > Regex Bugs:
> > > > > 
> > > > > * The Regex does not allow:
> > > > >   
> > > > >   * Filenames starting with '.' -> ".gitignore"
> > > > >   * Filenames containing valid special characters like "testv2.xml"
> > > > >   * A few more but my PC crashed and I'm writing this the second time and am too lazy to test everything again
> > > > > 
> > > > > Is there maybe a method provided by the Java files interface, that checks if a given name is valid on an operating system? Who knows what will happen on linux distros now 😆
> > > > > Alternatively we can accept the fact that our file generation is always limited to this regex.
> > > > 
> > > > 
> > > > One way we could do it is to rely on the Java `Paths.get(path: String)` method. This throws an `InvalidPathException` if the path cannot be converted. This solution also allows really weird names that include _%,&,$,§_ but we could argue that if java is fine with that, why not ?!. On the other hand the problem is with _.gitignore_ that the explorer does not show hidden files i guess, so the user does not see these files.
> > > 
> > > 
> > > Not sure if I get this right, but is showing hidden files not a windows explorer setting?
> > 
> > 
> > @Jooshua as far as i can tell i can see the hidden files in my explorer (e.g. the .gitignore file) but not in the project explorer
> 
> Same for me. So then let's ignore hidden files / files starting with . and call it a bug if @KamilRosiak even wants to support this feature :)

in latest commit the user is allowed to create [dot]-files too. If you agree with that feel free to finish the review :D
